### DISCUSSION TITLE
GS has no autonomous way to remove industries

### DISCUSSION THREAD

**Discussion by andythenorth** [OpenTTD/OpenTTD#11705] [Features]
### Version of OpenTTD

13-ish

### Expected result

GS has one or more methods to remove an industry from the map, which are _entirely_ within the control of the GS.

### Actual result

There are multiple routes to remove industries, none are inherently reliable.

### Use GSTile.DemolishTile()

See #10383 - this method can't remove industries currently, which is a bug.  This issue might be resolve by resolving #10383.

DemolishTile() does work if the player has enabled magic bulldozer via the cheat menu.  GS has no way to enable / disable cheats.

### Side-channel signalling to grf industries

The GS can use GSIndustry.SetControlFlags() to signal intents to an industry.  These can be interpreted by grf industries, and used to trigger the industry closure mechanism (via monthly industry callback 35).  

Control flags are currently: `
enum | IndustryControlFlags {  INDCTL_NO_PRODUCTION_DECREASE,  INDCTL_NO_PRODUCTION_INCREASE,  INDCTL_NO_CL }`

There is no appropriate flag for 'closure requested' or 'closure required'.  However a GS and grf can collaborate to abuse one of the other flags.

We could add further flags.

### Forced closure of an industry

This would be via monthly industry callback 35 (or the random equivalent). We delegated to grf the right to refuse closure, and I'm not proposing changing that.

### Forced removal of all of a type from the map

This would primarily be of use when modifying the map at game start, but there might be other uses.

Assuming #10383 was resolved, industries could be removed by DemolishTile().  Mass removal could be done by creating a GSIndustryList() instance, and filtering for industry type, then valuating with the tile index for each industry instance. Then running a foreach loop and calling DemolishTile on each instance.  This would consume a command for every instance, and commands are limited / expensive.

A method to mass-remove all of a type might be useful.

### Steps to reproduce

I can provide a test GS, but it's not much use really except as evidence for the findings above.

If we have proposed solutions, I can write a GS to test them.

**Comment by KajiKaede** [comment]
Facing this issue with an in-development GameScript of mine where I'm intending for industry closures to be a feature managed by said GS. Works fine for the server I'm currently developing it for, but it's harder for me to make it accessible to the general public if a cheat is required to be enabled just for core functionality.

I'd like to try and add this feature to the API myself, but I'm not presently sure what the best approach would be. Especially as #10383 / #11707 seems as though it will not be resolved? Not sure if I should contribute to the discussion there.

**Comment by nielsmh** [comment]
One feature that has been requested in perhaps a few different ways, would be having a "months (or days, or equivalent realtime units) until industry closes" counter as a core game mechanic. It could be accompanied with a list of potential countermeasures against the closing, such as providing service to the industry, or a message that it cannot be prevented.

The current core mechanic of secondary industries that have a risk of closure if they are not serviced within five years (IIRC) of opening, and primary industries that close if they lower production down past the last step, could possibly be converted to this mechanic, such that instead of immediately closing, they instead take on a countdown and offer the player(s) a way to save it.
NewGRF industries could possibly also have a new return value for the monthly production change callback (if there's any room for that) to indicate that they will close in X months, as an alternative to the immediate closure. They would also need a way to reset/clear the countdown.

Finally, GS could have a way to set a countdown on an industry, together with a string explaining how to save it. If no "how to save" string is given, the game can use a built-in string indicating that it's not possible to save the industry.

If GS sets a closure countdown on an industry, and the industry's own rules (game core or NewGRF) also sets a countdown, some kind of resolution needs to be offered. Probably keep track of both countdowns individually, and make it clear to players that they need to clear both of two independent conditions to save the industry.

**Reply by KajiKaede** [reply]
> instead of immediately closing, they instead take on a countdown and offer the player(s) a way to save it.

That's actually something I wanted to incorporate with the GameScript I've been cooking up, since GSGoal's question system seems flexible enough to allow for some fun experimentation there.

I do like the idea of incorporating a closing timer for industries with a potential goal to save them. Not sure how fit it would be seen for vanilla industries by default, but either way, I do think this sounds like a good idea.
